### ISSUE TITLE
KafkaConnect - Error retrieving Avro schema for id 54, Subject not found.; error code: 40401

### ISSUE THREAD

**Issue by JimFawkes** [aiven/karapace#258]
# TLDR
KafkaConnect S3 Sink Connector breaks for Avro Converter when Topic and Schema Name do not match and works if they do.

# What happened?
We are currently building a KafkaConnect POC with an S3 Sink. For this we setup the `Confluent - Amazon AWS S3 sink` connector (the `Aiven - Amazon AWS S3 sink` connector is not working but also not showing any errors, therefore the Confluent Connector). 

Connector Configuration:

- `Confluent - Amazon AWS S3 sink`
- Version: 5.0.0
- Avro Converter
- Schema Registry Config (URL/Credentials)
- Topic: `my.topic`

General Info:
- Schema Name: `events.v1.my.topic.created_key`
- Schema ID 54
- Events Avro Serialized

With this config we got an error saying: 
```
[...]
Caused by: org.apache.kafka.connect.errors.DataException: my.topic
[...]
Caused by: org.apache.kafka.common.errors.SerializationException: Error retrieving Avro schema for id 54
Caused by: io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException: Subject not found.; error code: 40401
[...]
```
_(See full traceback below)_ 

After some debugging we got it to work for a different topic which has a schema name matching it:

Topic Name: `my.othertopic`
Schema Name: `my.othertopic-key`

**Note**: Producers and Consumers have no issue with not matching topic vs schema names.

# What did you expect to happen?
We expected it to work since for Kafka or the Schema Registry there is no requirement of matching Topic and Schema Name.

# What else do we need to know?
This seems to be related: https://github.com/confluentinc/schema-registry/issues/343

Include your platform, version, and any other information that seems relevant.

- Kafka Version: 2.7
- Cloud: Amazon Web Services: Ireland (eu-west-1)

## Connector Traceback
```
org.apache.kafka.connect.errors.ConnectException: Tolerance exceeded in error handler
    at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndHandleError(RetryWithToleranceOperator.java:206)
    at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execute(RetryWithToleranceOperator.java:132)
    at org.apache.kafka.connect.runtime.WorkerSinkTask.convertAndTransformRecord(WorkerSinkTask.java:495)
    at org.apache.kafka.connect.runtime.WorkerSinkTask.convertMessages(WorkerSinkTask.java:475)
    at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:325)
    at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:229)
    at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:201)
    at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:185)
    at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:235)
    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.apache.kafka.connect.errors.DataException: my.topic
    at io.confluent.connect.avro.AvroConverter.toConnectData(AvroConverter.java:98)
    at org.apache.kafka.connect.storage.Converter.toConnectData(Converter.java:87)
    at org.apache.kafka.connect.runtime.WorkerSinkTask.lambda$convertAndTransformRecord$0(WorkerSinkTask.java:495)
    at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndRetry(RetryWithToleranceOperator.java:156)
    at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndHandleError(RetryWithToleranceOperator.java:190)
    ... 13 more
Caused by: org.apache.kafka.common.errors.SerializationException: Error retrieving Avro schema for id 54
Caused by: io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException: Subject not found.; error code: 40401
    at io.confluent.kafka.schemaregistry.client.rest.RestService.sendHttpRequest(RestService.java:209)
    at io.confluent.kafka.schemaregistry.client.rest.RestService.httpRequest(RestService.java:235)
    at io.confluent.kafka.schemaregistry.client.rest.RestService.lookUpSubjectVersion(RestService.java:302)
    at io.confluent.kafka.schemaregistry.client.rest.RestService.lookUpSubjectVersion(RestService.java:290)
    at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getVersionFromRegistry(CachedSchemaRegistryClient.java:130)
    at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getVersion(CachedSchemaRegistryClient.java:231)
    at io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer.schemaVersion(AbstractKafkaAvroDeserializer.java:202)
    at io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer.deserialize(AbstractKafkaAvroDeserializer.java:158)
    at io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer.deserializeWithSchemaAndVersion(AbstractKafkaAvroDeserializer.java:243)
    at io.confluent.connect.avro.AvroConverter$Deserializer.deserialize(AvroConverter.java:134)
    at io.confluent.connect.avro.AvroConverter.toConnectData(AvroConverter.java:85)
    at org.apache.kafka.connect.storage.Converter.toConnectData(Converter.java:87)
    at org.apache.kafka.connect.runtime.WorkerSinkTask.lambda$convertAndTransformRecord$0(WorkerSinkTask.java:495)
    at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndRetry(RetryWithToleranceOperator.java:156)
    at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndHandleError(RetryWithToleranceOperator.java:190)
    at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execute(RetryWithToleranceOperator.java:132)
    at org.apache.kafka.connect.runtime.WorkerSinkTask.convertAndTransformRecord(WorkerSinkTask.java:495)
    at org.apache.kafka.connect.runtime.WorkerSinkTask.convertMessages(WorkerSinkTask.java:475)
    at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:325)
    at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:229)
    at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:201)
    at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:185)
    at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:235)
    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by hackaugusto** [comment]
> We expected it to work since for Kafka or the Schema Registry there is no requirement of matching Topic and Schema Name.

That is not really correct. The converter needs a way to find a matching subject, and that is by default based derived from the topic's name:

Kafka connect needs a [`key.converter`](https://kafka.apache.org/documentation/#connectconfigs_key.converter) and [`value.converter`](https://kafka.apache.org/documentation/#connectconfigs_value.converter) to define the data format in a topic.

From the [Kafka connect docs](https://kafka.apache.org/documentation.html#connect_user):

> key.converter - Converter class used to convert between Kafka Connect format and the serialized form that is written to Kafka. This controls the format of the keys in messages written to or read from Kafka, and since this is independent of connectors it allows any connector to work with any serialization format. Examples of common formats include JSON and Avro.
> 
> value.converter - Converter class used to convert between Kafka Connect format and the serialized form that is written to Kafka. This controls the format of the values in messages written to or read from Kafka, and since this is independent of connectors it allows any connector to work with any serialization format. Examples of common formats include JSON and Avro.

These settings can be kept or overwritten on a connector basis.

In your case you're using [`io.confluent.connect.avro.AvroConverter` as a converter](https://www.confluent.io/hub/confluentinc/kafka-connect-avro-converter), which integrates with Schema Registry. The converter has a mechanism to associate a topic's key with a subject name, done through `key.subject.name.strategy`, and `value.subject.name.strategy` for the value. From the [converter docs](https://docs.confluent.io/platform/current/schema-registry/connect.html#sr-connect-config-options):

> key.subject.name.strategy
>
>    Determines how to construct the subject name under which the key schema is registered with Schema Registry. For additional information, see Schema Registry Subject Name Strategy.

and:

> By default, <topic>-key is used as subject

[Here you can find additional details](https://docs.confluent.io/platform/current/schema-registry/serdes-develop/index.html#sr-schemas-subject-name-strategy). I hope that helps.

**Comment by JimFawkes** [comment]
Thank you very much @hackaugusto. This was very helpful and I learned quite a bit.

Do you know if it is possible to define a custom Subject strategy? We are prefixing our schema names and not our topics.
Also, do you have an idea of why Producers/Consumers do not have an issue with this "custom strategy" we are using (we do not specify the schema name in any way).

For context, our topics look like `my.topic` while the corresponding schema name is: `events.v1.my.topic_record_name`

**Comment by hackaugusto** [comment]
> Do you know if it is possible to define a custom Subject strategy?

Yes, `value.subject.name.strategy` and `key.subject.name.strategy` are the names of the configurations options. The comment above has links to the docs with additional details.

> We are prefixing our schema names and not our topics.

I'm not sure if I follow. The topic name is not really changed, it is the name of the subject used to store schemas that is changed. 

> Also, do you have an idea of why Producers/Consumers do not have an issue with this "custom strategy" we are using (we do not specify the schema name in any way).

Well, the default strategy is to use the suffixes `-key` and `-value`. So if your subject name is `events.v1.my.topic_record_name` then there should be some customization going on.

Another possibility is that we are talking about different things :sweat_smile:
